2.x Update timber/twig/… filters

gchtr opened the pull request:
**Tickets**: #2408, #2284.

## Issue

While working on the documentation for #2408, I figured that the new filters to control what default filters and functions are added can also be used to add my own function to.

Actually, I find that it’s even a better way to add you own function, because the syntax is a little easier.

However, if I add my functions to `timber/twig/default_functions`, it feels a little weird, because of the name `default_functions`. As soon as I change something, it’s not the default anymore.

But what would be a better name? Of course: `timber/twig/functions`. But, that is actually a filter I would deprecate 🤪. But, what if we didn’t deprecate it, but change it to accept the functions array and not a Twig environment?

## Solution

This pull request shows you how it could be done, using a function that checks whether the passed argument is an instance of `Twig\Environment` with a priority of `9`.

Also

- Adds documentation for filters.

@nlemoine What do you think of this?

## Impact

If anyone used the following filters, they will break:

- `timber/twig/filters`
- `timber/twig/functions`
- `timber/twig/escapers`

Because this is a big breaking change, it would require us to include this into 2.0.

## Usage Changes

Upgrading is quite simple. Use `timber/twig` only for all things involving a Twig environment.

## Considerations

- Are there better names than `timber/twig/filters` and `timber/twig/filters`?
- We could use a similar pattern for escapers.

## Testing

I updated some tests.

## Todo

- [ ] ~Add more deprecation tests.~
- [x] Update Guides
- [x] Update Upgrade Guide

nlemoine commented:
Look good to me! 👍

acobster commented on lib/Twig.php:
```diff
@@ -340,13 +443,12 @@ public function get_timber_filters() {
 					$name,
 					$function['callable'],
-					isset($function['options']) ? $function['options'] : []
+					isset( $function['options'] ) ? $function['options'] : []
```
same opportunity here for null-coalescing.

acobster reviewed (commented):
Looks great, just a few minor style nitpicks.

jarednova commented:
FYI: the test failures here are unrelated to the PR (an issue with what themes WP can test against). These have been fixed in both the `2.x` and `master` branches, so you can ignore Travis failures related to those. @acobster please let me know if you don't have sufficient merge privileges to override the Travis check once this is ready.

gchtr commented on lib/Twig.php:
```diff
@@ -199,13 +266,12 @@ public function get_timber_functions() {
 					$name,
 					$function['callable'],
-					isset($function['options']) ? $function['options'] : []
+					isset( $function['options'] ) ? $function['options'] : []
```
Thanks for spotting that one 👍. Done.

gchtr commented on lib/Twig.php:
```diff
@@ -188,7 +217,45 @@ public function get_timber_functions() {
+		 *
+		 *     // Remove a function.
+		 *     unset( $functions['bloginfo'] );
+		 **
```
I never would have seen those. Thanks! 👍

gchtr commented on lib/Twig.php:
```diff
@@ -328,8 +394,45 @@ public function get_timber_filters() {
+		 *
+		 *     // Remove a filter.
+		 *     unset( $filters['list'] );
+		 **
```
Done

gchtr commented:
Okay, I worked some more on this.

I created a new Extending Twig Guide in [`ba6d10f`](https://github.com/timber/timber/pull/2419/commits/ba6d10f3cf34f3cffaabb29e422bf2c6732064a2) that adds proper documentation for the filters and merges some documentation that was scattered around different places into it. (I figured there’s a lot of other improvements I want to make to the Twig documentation in general, but that will come later.)

And I realized that it’s not really possible to add a proper deprecation for this, because obviously you can’t use `addFilter()` or `addFunction()` if `$twig` is an array. The deprecation will never run, because PHP will break before that.

That’s why I removed the non-working deprecation handling in [`3ed1b77`](https://github.com/timber/timber/pull/2419/commits/3ed1b77b203f40f404bc93a058eff1090d2bde59). Instead, I updated the Upgrade Guide with some helpful error message text in [`083dfa5`](https://github.com/timber/timber/pull/2419/commits/083dfa5963889e10a167f0db510fcefdc5140636) that developers will hopefully find if they search for that error.

@jarednova I’d be glad if you could have a look at this.

jarednova commented:
Thanks for catching this up with the rest of `2.x` @gchtr! It looks like we've got two failing tests with ...

```
 TestTimberTwigObjects::testTimberTermsInTwig
TestTimberTwigObjects::testTermsInTwig
```

... are those expected failures based upon the issues you flagged in your comment from 5/31? Or are those unrelated and we should hunt down what's triggering the fails?

gchtr commented:
> Thanks for catching this up with the rest of `2.x` @gchtr! It looks like we've got two failing tests with ...
> 
> ```
>  TestTimberTwigObjects::testTimberTermsInTwig
> TestTimberTwigObjects::testTermsInTwig
> ```
> 
> ... are those expected failures based upon the issues you flagged in your comment from 5/31? Or are those unrelated and we should hunt down what's triggering the fails?

Apparently, I didn’t update against the current 2.x branch yet. Now it looks fine and it’s ready for merge.

jarednova reviewed (approved):
Good catch on the upgrade guide!
